### PR TITLE
[v7r2] reuse Operations class for other section

### PR DESCRIPTION
**Pull request by TaykYoku**
BEGINRELEASENOTES
This small modification makes it possible to use the search logic in setups and VO paths that realised in the `Operations` for helper classes of other sections.

*Configuration
CHANGE: add possibility to reuse Operations class for other section

ENDRELEASENOTES

### CODE REVIEW

**Comment by TaykYoku** [comment]
The `pilot wrapper / PilotWrapper (2.7.5) (pull_request)` test failed with:
```
Trying https://diracproject.web.cern.ch/diracproject/tars/Pilot/DIRAC/master/
2021-07-09 12:14:21 UTC ERROR    https://diracproject.web.cern.ch/diracproject/tars/Pilot/DIRAC/master/ unreacheable (this is normal!)
```
And this, IMHO, does not correspond to the message "this is normal!".

File: src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py [hunk 1]
```diff
@@ -67,7 +67,38 @@
...
 
     3. if no VO nor group is provided, the VO will be guessed from the proxy,
-    but this works iff the object is instantiated by a proxy (and not, e.g., using a server certificate)
+    but this works if the object is instantiated by a proxy (and not, e.g., using a server certificate)
+
+    You can also specify the main section::
+
+      Operations(vo=VOName, mainSection='/mainSectionName').getValue('someSection/someOption')
+
+    in this case, the configuration will be searched in the following sections with the name 'mainSectionName'::
```
Comment (fstagni) [inline src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py:76]: I am not sure I understand. If there are many sections named "mainSectionName", and in each of them there's a "someSection/someOption", what's returned?

Comment (TaykYoku) [reply src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py]: I updated the documentation to more details.

The merger takes place in the following order (which is almost no different from the Operations usual):
- /`mainSectionName`/...
- /Operations/`setup`/`mainSectionName`/...
- /Operations/`vo`/Defaults/`mainSectionName`/...
- /Operations/`vo`/`setup`/`mainSectionName`/...

File: src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py [hunk 2]
```diff
@@ -89,63 +134,90 @@ class Operations(object):
...
+
+    self.__paths = [os.path.join('/', mainSection)] if mainSection else [os.path.join(basePath, 'Defaults')]
+    if self._setup:
+      self.__paths.append(os.path.join(basePath, self._setup, mainSection))
+    if self._vo:
+      self.__paths.append(os.path.join(basePath, self._vo, 'Defaults', mainSection))
+      if self._setup:
+        self.__paths.append(os.path.join(basePath, self._vo, self._setup, mainSection))
+
+  def _cacheExpired(self):
```
Comment (fstagni) [inline src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py:195]: Is this ever used?

Comment (TaykYoku) [reply src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py:195]: I plan to use it [here](https://github.com/DIRACGrid/WebAppDIRAC/pull/498/files#diff-6e1076bea0234c7ab02f6ed366399a69eab4e8588cb5f6214e3c62a042468df0), and generally think it's a useful method

File: src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py [hunk 3]
```diff
@@ -67,7 +67,51 @@
...
+                      anotherOption = anotherValueInProductionForVO
+                      mySection/
+                          myOption = myValueInProductionForVO
+              Certification/              -- this section will not be considered because Production setup is specified
+                  mainSection/
+                      anotherOption = anotherValueInCertificationForVO
+
+    section option will be merged in the following order::
+
+      /mainSection/mySection/myOption -- set a "myValue" for a myOption
```
Comment (fstagni) [inline src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py:109]: "set" or "get"? (also subsequent ones)

Comment (fstagni) [reply src/DIRAC/ConfigurationSystem/Client/Helpers/Operations.py:109]: Ah OK thanks.

**Review by fstagni — APPROVED** [review]

**Comment by chaen** [comment]
Honestly, I do not think it is a good idea to do it this way. The `Helper/Operations` has to search ONLY in `Operations` section. However, it is a very good idea to generalize the logic. So why not having a `BaseHelper` implementing the logic, and the `Operations` helper using this logic on the `Operations` section, etc ?

**Comment by TaykYoku** [comment]
I don't know the history of Operations creation, but having got acquainted with structure it seems that this section is just specially created for distribution of a configuration between setups and VOs

**Comment by TaykYoku** [comment]
I don't know the history of Operations creation, but having got acquainted with structure it seems that this section is just specially created for distribution of a configuration between setups and VO.

**Comment by atsareg** [comment]
I tend to agree with Chris. The proposed implementation can be renamed as, for example, VOConfigHelper. Then the Operations helper will be the instantiation with the /Operations as its mainSection (I would better call it rootSection) and WebApp helper correspondingly with WebApp as rootSection. 
Right now there is no complete symmetry between Operations and WebApp structure in the CS. There is no WebApp/<Setup> level. I think this should be better introduced. Yes, the CS will have to be changed by copying the contents of WebApp to WebApp/Defaults, but the overall structure will become more elegant.

**Comment by atsareg** [comment]
After a bit more thinking and discussion, I come to a conclusion that the implemented logic is correct. The Operations section is the only section which structured and used such that the Setup/VO specific information is properly discovered and combined with a general information. We do not introduce here a possibility of new sections that can be structured and interpreted in this way. So, let's the Operations() helper only behave on this section and no new base class is needed. 
The only modification to the Operations helper is to allow it to consider default parameters also from some arbitrary section. This is a minor change which provides convenience and backward compatibility if we want to add some more Setup/VO specific information into the Operations section but we keep defaults in the original section and do not move them to /Operations/Defaults. I think that this solution is quite elegant, not intrusive and completely backward compatible.

**Review by atsareg — APPROVED** [review]

**Comment by chaen** [comment]
then I am afraid I do not understand. Can you maybe give examples of what this does ? (And of course update the documentation accordingly :-) )

**Comment by TaykYoku** [comment]
The logic described in the old documentation should work as before. I have added documentation for use in the context of a some root section. I also tried to display all possible situations in `Test_Operations`(can be used as an example).
I do not quite understand what information is missing? Is the documentation incomprehensible in general and needs to be rewritten?
There may be a specific configuration example that is worth considering.

**Comment by atsareg** [comment]
I give you an example. We have a WebApp section in the CS which so far was used without the need to store VO specific info. Now we have such a need, so the obvious solution is to have a WebApp subsection in a corresponding /Operations section. But we want to keep the default WebApp parameters as they are without affecting current installations. Now this will be achieved in the code by simply specifying an (optional) mainSection argument.

**Comment by andresailer** [comment]
Why not _also_ search `/Operations/Defaults/<mainSection>`? Then one can chose to move WebApp to `/Operations/Defaults/` for consistency with other VO or Setup specific config.

**Comment by TaykYoku** [comment]
there is a nuance here, I think we can say that there are two types of requests:
- general queries, such as `gConfig.getValue('/WebApp/Port')`. That is, when a general configuration not specific to the VO is requested.
- session queries that are made in the context of some VO, for example. In this case, it is convenient to describe VO-specific things in the **Operations** section and use `Operations(vo=myVO, mainSection='/WebApp').getValue('Title')` for example.

If you look at it from this angle, I don't see the point in considering **/Operations/Defaults** when we already have the main section in the root as defaults.
IMHO the use of **/Operations/Defaults** is justified when we store some options placed arbitrarily, without being tied to the structure of sections in the root.

**Comment by andresailer** [comment]
Well, but look at it from the consistency angle, then you must also look in `/Operations/Defaults`. I really do not want to remember special cases for historical reasons. IMHO `/WebApp` should just be deprecated.

**Comment by TaykYoku** [comment]
`WebApp` its just example, we can say the same for `Resources` for example.

**Comment by TaykYoku** [comment]
OK, in general, I have no arguments to oppose the desire, I just explain my vision. I will make changes according to how all of you decides

**Comment by chaen** [comment]
I also think that consistency is paramount. One of the big issues we have with DIRAC (be it developers, users, administrators) is that there are always different ways to do the same things, and always specific cases, mostly for historical reasons. I would be in favor to aim at consistency, so @andresailer 's proposal to remove `WebApp` section is what makes the most sense to me. 

And @TaykYoku you are right ! It is very likely that `Resources` should follow the same path eventually :-)

**Comment by TaykYoku** [comment]
OK, I'll try to summarize raised issues:
- Is the documentation sufficient or needs to be supplemented or changed?
- Do I need to use `/Operations/Defaults` as the second default directory when I use `mainSection` as the default?
- where is the place for sections that may contain data that depend on VO and setups (WebApp, Resources and others) and how to organize them? IMHO this should be a separate issue, because it seems to affect a large part of the configuration.

**Comment by chaen** [comment]
What I think should be done is:

* Move `WebApp` into `Operations/Defaults/`
* Let the `Operations` helper be for the time being
* decide how we want to move forwards for declaring `Resources` in the context of multi VO. I suspect that having a structure similar to `Operations` is potentially a good idea. But that requires discussion first
* If `Resources` is to be structured like `Operations`, factorize the logic into this `BaseHelper` class (we will see for a name later :-) )  and just use it for `Resources` and `Operations`

But all this is too much meat for this PR, and of course will require some transitions over minor releases (if not major)

**Comment by TaykYoku** [comment]
@chaen agree that is too much meat for this PR. I started preparing the RFC for discussion in this context

**Comment by fstagni** [comment]
I have the feeling that this PR could be merged without (big) modifications. But the previous comment should then become an issue.

**Comment by chaen** [comment]
As agreed during the BILD, this can be closed, as it will be completely replaced with the outcome of https://github.com/DIRACGrid/DIRAC/issues/5287

**Pull request closed without merging**